### PR TITLE
feat: add PodDecoration controller

### PR DESCRIPTION
**Pull request by Eikykun**
#### 1. Does this PR affect any open issues?(Y/N) and add issue references

- [ ] N
- [x] Y,   #103  

#### 2. What is the scope of this PR (e.g. component or file name):
new api apis/apps/v1alpha1/poddecoration_types.go
pkg/controllers/collaset
pkg/controllers/poddecoration
poddecoration webhook

#### 3. Provide a description of the PR(e.g. more details, effects, motivations or doc link):

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Other

#### 4. Are there any breaking changes?(Y/N) and describe the breaking changes(e.g. more details, motivations or doc link):

- [x] N
- [ ] Y 

#### 5. Are there test cases for these changes?(Y/N) select and add more details, references or doc links:

- [x] Unit test
- [ ] Integration test
- [ ] Benchmark (add benchmark stats below)
- [ ] Manual test (add detailed scripts or steps below)
- [ ] Other

#### 6. Release note

Please refer to [Release Notes Language Style Guide](https://kusionstack.io/docs/governance/release-policy/) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/controllers/poddecoration/poddecoration_controller.go [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+		pods := affectedPods[collaSet.Name]
+		detail := appsv1alpha1.PodDecorationWorkloadDetail{
+			AffectedReplicas: int32(len(pods)),
+			CollaSet:         collaSet.Name,
+		}
+		status.MatchedPods += int32(len(pods))
+		for _, pod := range pods {
+			currentRevision := utilspoddecoration.GetDecorationGroupRevisionInfo(pod).
+				GetGroupPDRevision(instance.Spec.InjectStrategy.Group, instance.Name)
+			podInfo := appsv1alpha1.PodDecorationPodInfo{
```
Comment (wu8685) [inline pkg/controllers/poddecoration/poddecoration_controller.go:193]: put podInfo initialization into `if !disablePodDetail` statement?

File: pkg/controllers/poddecoration/poddecoration_controller.go [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+				podInfo.IsNotInjected = true
+			}
+			if !disablePodDetail {
+				detail.Pods = append(detail.Pods, podInfo)
+			}
+		}
+		details = append(details, detail)
+	}
+	fullControlByOthPD := heaviest != nil && heaviest.Name != instance.Name && heaviest.Status.CurrentRevision != ""
+	status.IsEffective = BoolPoint(instance.DeletionTimestamp == nil && (!fullControlByOthPD || hasEffectivePods))
```
Comment (wu8685) [inline pkg/controllers/poddecoration/poddecoration_controller.go:219]: BoolPointer?

File: pkg/controllers/poddecoration/poddecoration_controller.go [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+		}
+	}
+	for key, pods := range affectedPods {
+		sort.Slice(pods, func(i, j int) bool {
+			return pods[i].Name < pods[j].Name
+		})
+		affectedPods[key] = pods
+	}
+	collaSetList := &appsv1alpha1.CollaSetList{}
+	if err = r.List(ctx, collaSetList, &client.ListOptions{Namespace: instance.Namespace}); err != nil {
```
[the remark this answers is not in the document]
Comment (Eikykun) [reply pkg/controllers/poddecoration/poddecoration_controller.go:283]: We may need to select the collaset without pods.

File: pkg/controllers/collaset/synccontrol/sync_control.go [hunk 4]
```diff
@@ -348,37 +390,41 @@ func extractAvailableContexts(diff int, ownedIDs map[int]*appsv1alpha1.ContextDe
...
 
 	// 2. decide Pod update candidates
 	podToUpdate := decidePodToUpdate(cls, podUpdateInfos)
 
 	// 3. prepare Pods to begin PodOpsLifecycle
 	podCh := make(chan *PodUpdateInfo, len(podToUpdate))
-	for _, podInfo := range podToUpdate {
-		if podInfo.IsUpdatedRevision {
+	for i, podInfo := range podToUpdate {
+		if podInfo.IsUpdatedRevision && !podInfo.PodDecorationChanged {
```
Comment (wu8685) [inline pkg/controllers/collaset/synccontrol/sync_control.go:412]: We may move `!podInfo.PodDecorationChanged` into `decidePodToUpdate(cls, podUpdateInfos)` to decide, together with CollaSet, which Pod need to update.

File: pkg/controllers/collaset/synccontrol/update.go [hunk 5]
```diff
@@ -234,7 +291,11 @@ func (u *InPlaceIfPossibleUpdater) AnalyseAndGetUpdatedPod(cls *appsv1alpha1.Col
 	return
 }
 
-func (u *InPlaceIfPossibleUpdater) diffPod(currentPod, updatedPod *corev1.Pod) (inPlaceSetUpdateSupport bool, onlyMetadataChanged bool) {
+func (u *inPlaceIfPossibleUpdater) patchPodDecorations() {
```
Comment (wu8685) [inline pkg/controllers/collaset/synccontrol/update.go:294]: Remove this unused method?

File: pkg/controllers/collaset/synccontrol/update.go [hunk 6]
```diff
@@ -164,28 +187,62 @@ type ContainerStatus struct {
...
+		notFound, currentPodDecorations, err = utilspoddecoration.GetPodDecorationsByPodAnno(u.ctx, u.Client, podUpdateInfo.Pod)
+
+		if err != nil {
+			return false, false, nil, err
+		}
+		// if NotFound PD, recreate pod.
+		if notFound {
+			return false, false, nil, err
+		}
+		if err = utilspoddecoration.PatchListOfDecorations(currentPod, currentPodDecorations); err != nil {
```
Comment (wu8685) [inline pkg/controllers/collaset/synccontrol/update.go:229]: Can we move this patch PD operation into method `collasetutils.NewPodFrom()`?

File: pkg/controllers/utils/poddecoration/sort.go [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+	podDecorations := BuildSortedPodDecorationPointList(pdList)
+	if len(podDecorations) > 0 {
+		return podDecorations[0], nil
+	}
+	return
+}
+
+func lessPD(a, b *appsv1alpha1.PodDecoration) bool {
+	if a.Spec.InjectStrategy.Group == b.Spec.InjectStrategy.Group {
+		if *a.Spec.InjectStrategy.Weight == *b.Spec.InjectStrategy.Weight {
```
[the remark this answers is not in the document]
Comment (Eikykun) [reply pkg/controllers/utils/poddecoration/sort.go:78]: It will set default value by mutating webhook

**Review by wu8685 — APPROVED** [review]
lgtm

**Pull request merged**